tornado/locks.py

```diff
@@ -0,0 +1,62 @@
+__all__ = ['Condition']
+
+import collections
+
+from tornado import gen, ioloop
+from tornado.concurrent import Future
+
+
+class Condition(object):
+    """A condition allows one or more coroutines to wait until notified.
+
+    Like a standard Condition_, but does not need an underlying lock that
+    is acquired and released.
+
+    .. _Condition: http://docs.python.org/library/threading.html#threading.Condition
+    """
+
+    def __init__(self, io_loop=None):
+        self.io_loop = io_loop or ioloop.IOLoop.current()
+        self.waiters = collections.deque()  # Futures.
+
+    def __str__(self):
+        result = '<%s' % (self.__class__.__name__, )
+        if self.waiters:
+            result += ' waiters[%s]' % len(self.waiters)
+        return result + '>'
+
+    def wait(self, deadline=None):
+        """Wait for `.notify`. Returns a `.Future`.
+
+        Raises `.TimeoutError` if the condition is not notified before
+        ``timeout``, which may be specified in any form allowed by
+        `.IOLoop.add_timeout` (i.e. a `datetime.timedelta` or an absolute time
+        relative to `.IOLoop.time`)
+        """
+        waiter = Future()
+        self.waiters.append(waiter)
+        if deadline:
+            timed = gen.with_timeout(deadline, waiter, self.io_loop,
+                                     quiet_exceptions=gen.TimeoutError)
+
+            # Set waiter's exception after the deadline so notify(n) skips it.
+            gen.chain_future(timed, waiter)
```

Comment:
Import `chain_future` from its real location in `tornado.concurrent` instead of the indirect import through `gen`.

This is subtle; it's unfortunate that the timeout parameter to `wait()` behaves differently from wrapping the call in a with_timeout. I don't see a way around it though since you don't want to waste a `notify(1)` on a waiter that's not there if you can help it (although this implementation is less bulletproof than a `threading.Condition` since there is no guarantee that the caller will yield the returned Future immediately).

If timeouts are used and the `Condition` is infrequently notified, the old `Futures` will build up. Is it worthwhile to GC them? (`threading.Condition` allows spurious wakeups; assuming we do the same a cheap way to GC would simply be to `notify(1)` after every N timeouts.

`threading.Condition.wait` doesn't raise an exception when the timeout is reached (again because spurious wakeups are allowed). Prior to Python 3.2 you couldn't even tell whether the timeout had been reached; now it returns a bool.